### PR TITLE
Update Serverless Framework's AWS integration

### LINKED ISSUE #3284: [error] deploy aws serverless : @netlify/zip-it-and-ship-it is not installed
```
my system

CPU : i7-2600
OS : ubuntu 20
```

hello.
I'm trying to deploy by referring to the document ' https://redwoodjs.com/docs/deploy '.
when I type 'yarn rw deploy aws serverless --side api'  , I encountered the error

```
lkb@lkb-System-Product-Name:~/project/redwood-blog-2$ yarn rw deploy aws serverless --side api
yarn run v1.22.5
$ /home/lkb/project/redwood-blog-2/node_modules/.bin/rw deploy aws serverless --side api
  ❯ Checking pre-requisites
    ✔ Checking if Serverless framework is installed...
    ✖ Checking if @netlify/zip-it-and-ship-it is installed...
      → Looks like @netlify/zip-it-and-ship-it is not installed. Either run `yar
…
    Building and Packaging...
Command failed with exit code 1: yarn zip-it-and-ship-it --version
error Command "zip-it-and-ship-it" not found.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Looks like @netlify/zip-it-and-ship-it is not installed. Either run `yarn rw setup aws-serverless` or add it seperately as a dev dependency in the api workspace.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

so I run 
```
yarn rw setup aws-serverless
```
the result is 
```
lkb@lkb-System-Product-Name:~/project/redwood-blog-2$ yarn rw setup aws-serverless
yarn run v1.22.5
$ /home/lkb/project/redwood-blog-2/node_modules/.bin/rw setup aws-serverless
Done in 2.02s.
```

after setup, I restart 
```
yarn rw deploy aws serverless --side api
```
but error is the same. I think @netlify/zip-it-and-ship-it is not installed.
about this, Is there anyting I can try or Can I contribute something about it ?
thank you.

Also, It is my first github issue. so maybe I mistake something. please understand.

### PR DESCRIPTION
**Pull request by virtuoushub**
Work related to #3284 / #3686
---
- target latest node runtime
- improve CORS defaults
- remove deprecation warnings
- clean up packaging (unrelated prisma binaries not included) 

see also: https://github.com/redwoodjs/redwood/issues/3284#issuecomment-922375958

### CODE REVIEW

File: packages/cli/src/commands/setup/deploy/providers/aws-serverless.js
Comment (thedavidprice) [inline packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:28]: Ah, very nice. FYI: our build targets Nodejs 12, which shouldn't be a problem. But good to see 14 is available 'cause I'd like us to bump our internal transpile target in the future. (Not now, but likely once we approach Redwood v1 GA.

File: packages/cli/src/commands/setup/deploy/providers/aws-serverless.js
Comment (thedavidprice) [inline packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:19]: 🚀

File: packages/cli/src/commands/setup/deploy/providers/aws-serverless.js
Comment (virtuoushub) [reply packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:28]: Sound good, I will remove this change from the PR; is there another ticket where we might want to track the upgrade to 14?

Comment (thedavidprice) [reply packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:28]: Ah, I wasn't implying you need to remove it! Let's leave it for now but definitely confirm/review before finalizing this. 

I just meant to say:
1. it's good to see that AWS Lambdas has v14 support (it does, correct?)
2. this will help us force the conversation about if/when we bump our targets. And we also might need to document how the transpiration target can be set per-project's needs.

Comment (ajcwebdev) [reply packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:28]: You can check available runtimes [here](https://docs.aws.amazon.com/lambda/latest/dg/lambda-nodejs.html), but yes, Node 14 has been supported on [Lambda since February](https://aws.amazon.com/blogs/compute/node-js-14-x-runtime-now-available-in-aws-lambda/) and on [Lambda@Edge since April](https://aws.amazon.com/about-aws/whats-new/2021/04/lambda-edge-support-node14/). There's an [open issue for Node 16](https://github.com/aws/aws-lambda-base-images/issues/14) but wouldn't count on that coming out anytime soon.

File: packages/cli/src/commands/setup/deploy/providers/aws-serverless.js
Comment (virtuoushub) [reply packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:19]: Unfortunately, use of this plugin might not be what we want. We currently use [`httpApi` events](https://www.serverless.com/blog/aws-http-api-support) and not `http`; which this [plugin expects.](https://github.com/Droplr/serverless-api-cloudfront#error-troubleshooting) The plugin also does not seem to be updated much, so more of a roll our own approach seems warranted. 

---

Shout out to @ajcwebdev for helping smoke testing this.

Comment (thedavidprice) [reply packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:19]: Ah, bummer. I'm trying to get some help from folks at Serverless. Let's keep this top of the list for them to look at.

File: packages/cli/src/commands/setup/deploy/providers/aws-serverless.js
Comment (Irev-Dev) [inline packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:31]: `cors: true` is serverless shorthand for wild card origin and a bunch of headers, the biggest problem is the wildcard as that it's not allow by browsers for requests with credentials. Here's my suggestion for what we should add instead.

```suggestion
    cors:
      allowedOrigins:
        - '*' # the * wildcard will need be replaced with your domain to allow requests with credentials.
      allowCredentials: true
      allowedHeaders:
        - authorization
        - auth-provider
        - content-Type
```

Though we could also add these common aws headers that could be useful depending on what other aws services a Redwood user might employ.
```
              - X-Amz-Date
              - X-Api-Key
              - X-Amz-Security-Token
              - X-Amz-User-Agent
```

File: packages/cli/src/commands/setup/deploy/providers/aws-serverless.js
Comment (Irev-Dev) [inline packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:18]: dot env files are going to be included by default in the next serverless version, but the functionality can be opted into now, plus helps remove a dependency.

```suggestion
useDotenv: true

plugins:
```

if this suggestion is accepted it also needs to be removed from the setup command install, see:

```javascript
// packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:87
export const apiDevPackages = [
  '@netlify/zip-it-and-ship-it',
  'serverless-dotenv-plugin', // should be removed too!
]
```

Comment (Irev-Dev) [reply packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:18]: Note that in #3782 I added a [stage](https://github.com/redwoodjs/redwood/pull/3782/files#diff-f18dd32c39f000ab0ad47befa798fbf3c534c2f963f75be5ae82e0fc61a0dc30R38-R43) argument. Related to this as when deploying serverless will look for a `.env.{stage}` file before falling back to the `.env` file if non exists.

Comment (mnapoli) [reply packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:18]: > dot env files are going to be included by default in the next serverless version

Hi, quick note: `useDotenv: true` will work in Serverless Framework v2 and v3 (future next major) in the same way.

That means that you will need to keep `useDotenv: true` in serverless.yml if you want `.env` files to be loaded.

Note that the `serverless-dotenv-plugin` plugin works slightly differently from the built-in feature, so pick the one you prefer based on your requirements.

File: packages/cli/src/commands/setup/deploy/providers/aws-serverless.js
Comment (mnapoli) [reply packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:19]: Hi! Serverless Framework team member here 👋 

Could you clarify why CloudFront is used? Is it used:

- as a CDN for static assets/static website hosting (i.e. in front of S3)
- or as a CDN/cache for HTTP responses returned by Lambda (i.e. in front of API Gateway)

From what I understand it seems to be the second case, but I just want to make sure.

AFAIK here are your option to set up CloudFront with API Gateway in serverless.yml:

- deploy and configure CloudFront via CloudFormation code in `serverless.yml` (in the `resources` section) -> you should expect ~100 lines of YAML (I've done this a lot, this can be painful) -> I can help you with that if you choose to go that way
- or use a plugin, but AFAIK there are not many options - I've worked on a plugin that may help here: https://github.com/getlift/lift

Lift offers several "constructs" to deploy high-level components. One of them is "server-side website": that construct deploys a CloudFront distribution that serves responses from API Gateway.

You can read more about it here: https://github.com/getlift/lift/blob/master/docs/server-side-website.md

One piece of caution: that construct was built with backend frameworks in mind, like Laravel, Ruby on Rails, Django, Symfony… Not really for server-side rendering of frontend websites (even though one user [mentioned that it seemed to work fine too for these use-cases](https://github.com/getlift/lift/discussions/43#discussioncomment-896386)). From what I've read about Redwood, it's the GraphQL API running on Lambda? If that's the case, then it could work well here.

Also that leads me to one question: why use CloudFront in front of the GraphQL API?

By the way we also have one for static websites served with CloudFront + S3: https://github.com/getlift/lift/blob/master/docs/static-website.md (but it may not be what you need here).

Comment (thedavidprice) [reply packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:19]: Hi @mnapoli Thanks for jumping in here. Redwood has two considerations during deployment:
1. API functions
2. Web assets

We use a Jamstack-like deployment flow, which you can read about in the first two sections of this Doc:
- https://redwoodjs.com/docs/deploy

Our goal here is to deploy:
1. API Functions --> AWS Lambdas
2. Web assets --> CDN

The latter should be Cloudfront, correct?

## Bonus Time: Caching
Lastly, Redwood's GraphQL API Function is "cache-ready" out of the box: either mem or DB (eg. Redis). For serverless deployments, we normally reach for Redis, which might not be the only option on AWS Lambdas.

It's not a hard requirement for this PR, but it would 100% be something that gets people excited to use this Serverless setup. Any immediate reactions/ideas?

Comment (mnapoli) [reply packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:19]: Thanks, that's clearer. I see 2 scenarios:

Scenario A (simplest):

- website.com -> CloudFront -> S3
- api.website.com (i.e. another domain) -> API Gateway -> Lambda

In that case, [Lift's static website](https://github.com/getlift/lift/blob/master/docs/static-website.md) might be perfect here. If you don't want to use the Lift plugin, it is also perfectly possible to write the CloudFormation code manually (or [use "eject" to generate that code](https://github.com/getlift/lift#ejecting)). Note that Lift provides a few helpers as well to quickly deploy asset changes, and clear the CDN cache on deployment.

You get CDN for the static website, but CloudFront would not cover the API. One could use API Gateway's caching feature though (API Gateway has a caching system).

Scenario B:

- website.com/* -> CloudFront
    - /* -> S3
    - /api/* -> API Gateway -> Lambda

In other words, all requests go through CloudFront. Then CloudFront routes based on the path. CloudFront caching could then apply to web assets and API responses. Watch out: this scenario also involves extra CloudFront costs, because all API calls go through CloudFront.

In that case, you may need to write the CloudFront configuration manually.

File: packages/cli/src/commands/setup/deploy/providers/aws-serverless.js
Comment (mnapoli) [reply packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:18]: Usually, secrets are indeed better stored outside of the filesystem (because it makes sharing them across the team hard & risky, and there's always the risk of committing them/publishing them somewhere).

Common approaches with AWS (and Serverless Framework) are:

- store secrets in AWS SSM Parameter Store or AWS Secrets Manager, and reference them from serverless.yml (https://www.serverless.com/framework/docs/providers/aws/guide/variables/#reference-variables-using-aws-secrets-manager)
- use Serverless Dashboard parameters: https://www.serverless.com/framework/docs/guides/parameters

Comment (Irev-Dev) [reply packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:18]: Adding either of those options into the setup tutorial is a good idea.

Comment (Irev-Dev) [reply packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:18]: What's action should be taken from this conversation @thedavidprice?

I think maybe updating https://redwoodjs.com/docs/deploy#aws-serverless-deploy to include setting up the serverless dashboard with ci/cd and secrets/env-vars? should I make an issue? (and then resolve that issue 😝 )

**Comment by ajcwebdev** [comment]
Danny's notes, lots of good stuff we'll want to incorporate into this draft: https://github.com/redwoodjs/redwood/pull/3782#issuecomment-989144105

File: packages/cli/src/commands/setup/deploy/providers/aws-serverless.js
Comment (virtuoushub) [reply packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:19]: @mnapoli Lift's `static-website` construct is great, thanks for pointing me in this direction!

If we did decide to go with scenario B, do you know of a similar construct for spinning up that setup, or where I might start looking for CDK snippets/constructs? I'd like to avoid manual CloudFront if possible.

File: packages/cli/src/commands/setup/deploy/providers/aws-serverless.js
Comment (Irev-Dev) [reply packages/cli/src/commands/setup/deploy/providers/aws-serverless.js]: I'm not sure, but that sounds good if there is an option like that.

File: packages/cli/src/commands/setup/deploy/providers/aws-serverless.js
Comment (mnapoli) [reply packages/cli/src/commands/setup/deploy/providers/aws-serverless.js]: @virtuoushub For scenario B I'm wondering if https://github.com/getlift/lift/blob/master/docs/server-side-website.md could actually be a solution here 🤔 

Maybe a config like this would be worth testing:

```yaml
service: my-app
provider:
    name: aws

functions:
    function1:
        handler: home.handler
        events:
            -   httpApi: 'GET /.redwood/functions/function1/'
    # ...

constructs:
    website:
        type: server-side-website
        assets:
            '/index.html': dir-of-static-website/index.html
            '/css/*': dir-of-static-website/css
            '/js/*': dir-of-static-website/js

plugins:
    - serverless-lift
```

The downsides:

- the name of that construct (`server-side-website`) isn't really the best name here, but it's just a name 🤷 
- with that construct, since `/*` is directed by default to API URLs, you need to specify explicitly every root file in the `constructs.website.assets` section (you cannot do `'/*': dir-of-static-website/`, or at least I don't expect it to work)

**Comment by thedavidprice** [comment]
## Update
I’ve changed this PR to merge against a new branch:
- [serverless-deploy-3686](https://github.com/redwoodjs/redwood/tree/serverless-deploy-3686)

Also, moving and consolidating conversation to:
- #3686

**Comment by thedavidprice** [comment]
Please see discussion over here:
- https://github.com/redwoodjs/redwood/issues/3686#issuecomment-996926893

**Comment by dac09** [comment]
@virtuoushub do you think you could please update the title and description to explain the changes here? I can see what they are but when reviewing later on it might be helpful

**Comment by virtuoushub** [comment]
@dac09 / @jtoar re: @thedavidprice's [comment](https://github.com/redwoodjs/redwood/issues/3686#issuecomment-996926893), I'll defer to you on what we want to do with this PR.

File: packages/cli/src/commands/setup/deploy/providers/aws-serverless.js
Comment (dac09) [inline packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:53]: @virtuoushub could you give me a clue what this is required for? Everything else in this PR looks good.

Comment (virtuoushub) [reply packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:53]: @dac09 - they stem from reading through @thedavidprice's comment here: https://github.com/redwoodjs/redwood/issues/3284#issuecomment-922375958 and more specifically here as that first one is a rather large info dump: https://gist.github.com/thedavidprice/ac67656b1a3e2b32f31799cd4fb7cbec#gistcomment-3897679

I have not test this theory, but it is very possible #3782 does most of this cleanup for us already?

see also:
- https://www.prisma.io/docs/guides/deployment/deployment-guides/deploying-to-aws-lambda#package-pattern-in-serverlessyml
- https://gist.github.com/thedavidprice/ac67656b1a3e2b32f31799cd4fb7cbec#gistcomment-3897679

Comment (dac09) [reply packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:53]: I see makes sense. Let me take it for a spin, and will merge in if it deploys all ok!

File: packages/cli/src/commands/setup/deploy/providers/aws-serverless.js
Comment (dac09) [inline packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:32]: Hey @virtuoushub - I'm back :) 

I'm getting this error when trying to deploy, maybe there's something a little bit wrong? 

```
Serverless: DOTENV: Loading environment variables from .env:
Serverless: Configuration warning at 'provider.httpApi.cors': unrecognized property 'allowOrigins'
Serverless:
```

Comment (dac09) [reply packages/cli/src/commands/setup/deploy/providers/aws-serverless.js:28]: Nevermind, fixed it `allowOrigins` -> `allowedOrigins`

**Review by dac09 — APPROVED** [review]

**Pull request merged**
